File: 21-merge-two-sorted-lists/merge-two-sorted-lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} list1
 * @param {ListNode} list2
 * @return {ListNode}
 */
//  input: list1, list2, both sorted and linked lists
// output: 1 list sorted
// both lists are sorted 
// edge case: what if 1 list is longer than the other ?
// what if list is empty?
// [1,2,4]
// t
// [1,3]
var mergeTwoLists = function(list1, list2) {
    let dummy = new ListNode()
    let tail = dummy 
    while(list1 && list2) {
        if (list1.val < list2.val) {
            tail.next = list1
            tail = list1
            list1 = list1.next
        } else {
            tail.next = list2
            tail = list2
            list2 = list2.next
        }
    }
    if (list1) {
        tail.next = list1
    } else if (list2) {
        tail.next = list2
    }
    return dummy.next
}

File: 21-merge-two-sorted-lists/merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        # inputs: two sorted linked lists l1 and l2
        # outputs: head of merged sorted linked list

        if list1 is None and list2 is None:
            return list1
        if list1 is None and list2 is not None:
            return list2
        if list1 is not None and list2 is None:
            return list1
        
        if list1.val <= list2.val:
            list1.next = self.mergeTwoLists(list1.next, list2)
            return list1
        else:
            list2.next = self.mergeTwoLists(list1, list2.next)
            return list2
        